### ISSUE TITLE
modify KG2 neo4j phase of build system to build the synonymizer database?

### ISSUE THREAD

**Issue by saramsey** [RTXteam/RTX#1029]

**Comment by dkoslicki** [comment]
DK to modify `dumpdata.py` to point to local host

**Comment by dkoslicki** [comment]
SAR to provide links to S3 bucket `NodeSynonymizer.sqlite` files

**Comment by saramsey** [comment]
Idea is to have the KG2 build process updated to dump (as in build) the Synonymizer sqlite database and copy said sqlite database to the public S3 bucket.

**Comment by saramsey** [comment]
The RTX KG2 public S3 bucket has AWS URI: `s3://rtx-kg2-public/` and it is in the `us-west-2` region.

The RTX KG2 public S3 bucket has address:
[http://rtx-kg2-public.s3-website-us-west-2.amazonaws.com/](http://rtx-kg2-public.s3-website-us-west-2.amazonaws.com/)

which points to a static HTML file (an index file) `s3-index.html` that is under git version control.

We can certainly update the KG2 build system to produce an index HTML file for tracking versions of the synonymizer sqlite database files that are in the S3 bucket, and for providing HTTP URLs to said database files.

**Comment by amykglen** [comment]
not sure if we made it this far already, but one idea for the usage half of this new system could be:

* `config.json` has one `KG2` slot, somewhere in which the 'active' KG2 version is listed/accessible
* when code using `NodeSynonymizer` is run, `NodeSynonymizer` looks at that KG2 version number in the config file to figure out which node_synonymizer.sqlite file it should use (e.g., `node_synonymizer_KG2-1-0.sqlite` or `node_synonymizer_KG2-2-0.sqlite`)
* if that sqlite file doesn't already exist locally, `NodeSynonymizer` downloads it from S3

then if you want to use a new 'test' KG2 for dev work, you would just need to edit your local `config.json` to point to whatever KG2 version you want, and the rest would be taken care of.

and one idea to address the issue where `config.json` is downloaded excessively (overwriting your local changes) is to make it so that it's downloaded only if its last modified timestamp is more recent than the last modified timestamp of your local config file. (I use that method for updating the NGD database, [here](https://github.com/RTXteam/RTX/blob/master/code/ARAX/ARAXQuery/Overlay/compute_ngd.py#L266-L280).)

**Comment by ericawood** [comment]
I can modify tsv-to-neo4j.sh to build the SQLite database if that is desired.

**Comment by saramsey** [comment]
> I can modify tsv-to-neo4j.sh to build the SQLite database if that is desired.

Thanks, @ericawood .  If we decide to pull the trigger on this modification to the KG2 build system (and I think there will be a bit more discussion before we reach that stage), we'll probably want to have a new master shell script "build-kg2-neo4j.sh" that (1) calls "tsv-to-neo4j.sh", (2) calls a new shell script "make-synonymizer-db.sh" and (3) copies the sqlite file to the S3 bucket.

**Comment by saramsey** [comment]
> > I can modify tsv-to-neo4j.sh to build the SQLite database if that is desired.
> 
> Thanks, @ericawood . If we decide to pull the trigger on this modification to the KG2 build system (and I think there will be a bit more discussion before we reach that stage), we'll probably want to have a new master shell script "build-kg2-neo4j.sh" that (1) calls "tsv-to-neo4j.sh", (2) calls a new shell script "make-synonymizer-db.sh" and (3) copies the sqlite file to the S3 bucket.

It's possible some of these filenames will get changed, but you get the idea.

**Comment by saramsey** [comment]
> not sure if we made it this far already, but one idea for the usage half of this new system could be:
> 
>     * `config.json` has one `KG2` slot, somewhere in which the 'active' KG2 version is listed/accessible
> 
>     * when code using `NodeSynonymizer` is run, `NodeSynonymizer` looks at that KG2 version number in the config file to figure out which node_synonymizer.sqlite file it should use (e.g., `node_synonymizer_KG2-1-0.sqlite` or `node_synonymizer_KG2-2-0.sqlite`)
> 
>     * if that sqlite file doesn't already exist locally, `NodeSynonymizer` downloads it from S3
> 
> 
> then if you want to use a new 'test' KG2 for dev work, you would just need to edit your local `config.json` to point to whatever KG2 version you want, and the rest would be taken care of.
> 
> and one idea to address the issue where `config.json` is downloaded excessively (overwriting your local changes) is to make it so that it's downloaded only if its last modified timestamp is more recent than the last modified timestamp of your local config file. (I use that method for updating the NGD database, [here](https://github.com/RTXteam/RTX/blob/master/code/ARAX/ARAXQuery/Overlay/compute_ngd.py#L266-L280).)

I like that idea, with the addition that if the download is skipped due to the "local copy is newer" condition, a warning should be printed to stderr stating that a local config file is overriding the system-wide-default config file (if it is annoying, I suppose the warning could be made suppressible by a parameter to the `RTXConfiguration` constructor, perhaps).

**Comment by amykglen** [comment]
good idea!

**Comment by saramsey** [comment]
Are we still planning to do this?  It's been some months, so I wanted to check if this has already been worked-around, or if the issue is still pending.

**Comment by edeutsch** [comment]
I don't know myself. There are a whole series of things that should be triggered with a new KG2 that we plan to put into production.

**Comment by amykglen** [comment]
yeah, good question - I'm feeling less sure about the usefulness of this myself: given that new KG2 builds sometimes can break the node synonymizer, and that not every KG2 build is brought to ARAX production anyway, I'm not sure how helpful it would be to have the synonymizer build process automatically kicked off. especially since having a synonymizer isn't really useful for anyone until there's at least a KG2c built from it (also which often requires adjustments for a new KG2 version).. so maybe it's just something that should be paired with the KG2c build process.

**Comment by saramsey** [comment]
+1 for pairing with the KG2c build process. In practice we usually want to stop at the end of the KG2 build and do a bunch of inspection of build artifacts and reports, as well as testing in Neo4j, before we conclude whether or not the build was "successful".
